src/Sentry/Internal/CacheDirectoryCoordinator.cs

```diff
@@ -0,0 +1,153 @@
+    }
+
+    public bool TryAcquire()
+    {
+        if (_acquired)
+        {
+            return true;
+        }
+
+        lock (_gate)
+        {
+            if (_acquired)
+            {
+                return true;
+            }
+
+            if (_disposed)
+            {
+                return false;
+            }
+
+            try
+            {
+                _acquired = _fileSystem.TryCreateLockFile(_lockFilePath, out _lockStream);
+                return _acquired;
+            }
+            catch (Exception ex)
+            {
+                _logger?.LogDebug("Unable to acquire cache directory lock", ex);
+            }
+            finally
+            {
+                if (!_acquired && _lockStream is not null)
+                {
+                    try
+                    { _lockStream.Dispose(); }
+                    catch
+                    {
+                        // Ignore
+                    }
+                    _lockStream = null;
+                }
+            }
+
+            return false;
+        }
+    }
+
+    public void Dispose()
+    {
+        lock (_gate)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            _disposed = true;
+
+            if (_acquired)
+            {
+                try
+                {
+                    _lockStream?.Close();
+                }
+                catch (Exception ex)
+                {
+                    _logger?.LogError("Error releasing the cache directory file lock.", ex);
+                }
+            }
+
+            try
+            {
+                _lockStream?.Dispose();
+            }
+            catch (Exception ex)
+            {
+                _logger?.LogError("Error disposing cache lock stream.", ex);
+            }
+            _lockStream = null;
+        }
+    }
+}
+
+internal static class CacheDirectoryHelper
+{
+    public const string IsolatedCacheDirectoryPrefix = "isolated_";
+
+    internal static string? GetBaseCacheDirectoryPath(this SentryOptions options) =>
+        string.IsNullOrWhiteSpace(options.CacheDirectoryPath)
+            ? null
+            : Path.Combine(options.CacheDirectoryPath, "Sentry");
+
+    internal static string? GetIsolatedFolderName(this SentryOptions options)
+    {
+        var stringBuilder = new StringBuilder(IsolatedCacheDirectoryPrefix);
+#if IOS || ANDROID
+        // On iOS or Android the app is already sandboxed, so there's no risk of sending data to another Sentry's DSN.
+        // However, users may still initiate the SDK multiple times within the process, so we need an InitCounter
+        stringBuilder.Append(options.InitCounter.Count);
+#else
+        if (string.IsNullOrWhiteSpace(options.Dsn))
+        {
+            return null;
+        }
+        var processId = options.ProcessIdResolver.Invoke() ?? 0;
+        stringBuilder.AppendJoin('_', options.Dsn.GetHashString(), processId, options.InitCounter.Count);
```

Comment:
suggestion (nitpick):


```suggestion
        stringBuilder.AppendJoin('_', options.Dsn.GetHashString(), processId.ToString(), options.InitCounter.Count.ToString());
```

We're boxing both the `processId` and the `options.InitCounter.Count` from `int` to `object`.
We could `ToString` directly to avoid these boxing allocations.